javascript/extractor/src/com/semmle/js/dependencies/DependencyResolver.java

```diff
@@ -0,0 +1,221 @@
+package com.semmle.js.dependencies;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import com.google.gson.Gson;
+import com.semmle.js.dependencies.packument.PackageJson;
+import com.semmle.util.data.Pair;
+
+public class DependencyResolver {
+    private AsyncFetcher fetcher;
+    private List<Constraint> constraints = new ArrayList<>();
+
+    /** Packages we don't try to install because it is part of the same monorepo. */
+    private Set<String> packagesInRepo;
+
+    private static class Constraint {
+        final PackageJson targetPackage;
+        final SemVer targetPackageVersion;
+        final PackageJson demandingPackage;
+        final int depth;
+
+        Constraint(PackageJson targetPackage, SemVer targetPackageVersion, PackageJson demandingPackage, int depth) {
+            this.targetPackage = targetPackage;
+            this.targetPackageVersion = targetPackageVersion;
+            this.demandingPackage = demandingPackage;
+            this.depth = depth;
+        }
+
+        String getTargetPackageName() {
+            return targetPackage.getName(); // Must exist as you can't depend on a package without a name
+        }
+    }
+
+    public DependencyResolver(ExecutorService threadPool, Set<String> packagesInRepo) {
+        this.fetcher = new AsyncFetcher(threadPool, this::reportError);
+        this.packagesInRepo = packagesInRepo;
+    }
+
+    private void reportError(CompletionException ex) {
+        System.err.println(ex);
+    }
+
+    private void addConstraint(Constraint constraint) {
+        synchronized(constraints) {
+            constraints.add(constraint);
+        }
+    }
+
+    /**
+     * Returns the first version number mentioned in the given constraints, excluding upper bounds such as `< 2.0.0`,
+     * or `null` if no such version number was found.
```

Comment:
Would rather keep the current behaviour. An explicitly-mentioned version at least has a high likelihood of working, because it most likely worked at the point that constraint was added.